### PR TITLE
[query/ptypes] random function registry

### PR DESCRIPTION
**Pull request by akotlar**

### CODE REVIEW

**Comment by akotlar** [comment]
Not sure what this error is: 

deepest = 'HailException: block matrix must have at least one row'
full = 'is.hail.utils.HailException: block matrix must have at least one row\n\tat is.hail.utils.ErrorHandling$class.fatal(Er...ava:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.lang.Thread.run(Thread.java:748)\n\n\n'

    def deco(*args, **kwargs):
        import pyspark
        try:
            return f(*args, **kwargs)
        except py4j.protocol.Py4JJavaError as e:
            s = e.java_exception.toString()
    
            # py4j catches NoSuchElementExceptions to stop array iteration
            if s.startswith('java.util.NoSuchElementException'):
                raise
    
            tpl = Env.jutils().handleForPython(e.java_exception)
            deepest, full = tpl._1(), tpl._2()
            raise FatalError('%s\n\nJava stack trace:\n%s\n'
                             'Hail version: %s\n'
>                            'Error summary: %s' % (deepest, full, hail.__version__, deepest)) from None
E           hail.utils.java.FatalError: HailException: block matrix must have at least one row
E           
E           Java stack trace:
E           is.hail.utils.HailException: block matrix must have at least one row

File: hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 1]
```diff
@@ -453,7 +453,12 @@ abstract class RegistryFunctions {
 
       override val returnType: Type = rType
 
-      override def returnPType(argTypes: Seq[PType], returnType: Type): PType = if (pt == null) PType.canonical(returnType) else pt(argTypes)
+      override def returnPType(argPTypes: Seq[PType], returnType: Type): PType = {
+        val rt = if (pt == null) PType.canonical(returnType) else pt(argPTypes)
+
+        // applySeeded sets EmitCode.m missing if any argument not required
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: This is a misleading comment -- this requiredness is a property of the type, but `EmitCode.m` (missingness) is a property of the runtime value. Something like:
"the result of applySeeded is missing if any argument is missing" is correct.

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: What does "result" refer to here? The return value of applySeeded is an EmitCode

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: the emit code, yes

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: the emit code is missing (m is true) if any argument is missing

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: got it I think: the reader will know we're talking about the value present in the generated binary

File: hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 2]
```diff
@@ -453,7 +453,12 @@ abstract class RegistryFunctions {
 
       override val returnType: Type = rType
 
-      override def returnPType(argTypes: Seq[PType], returnType: Type): PType = if (pt == null) PType.canonical(returnType) else pt(argTypes)
+      override def returnPType(argPTypes: Seq[PType], returnType: Type): PType = {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: interesting, this change breaks test/hail/methods/test_statgen.py::Tests::test_ld_prune_no_prune - hail.utils.java.FatalError: HailException: block matrix must have at least one row

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: that's not good.

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: tracking it, here are the mnames and returnTypes leading up to the error:

```
returnType is int32 and mName is rand_cat
returnType is int32 and mName is rand_cat
returnType is float64 and mName is rand_unif
returnType is float64 and mName is rand_beta
returnType is int32 and mName is rand_cat
returnType is float64 and mName is rand_unif
returnType is float64 and mName is rand_beta
returnType is int32 and mName is rand_cat
returnType is int32 and mName is rand_cat
returnType is int32 and mName is rand_cat
returnType is float64 and mName is rand_unif
returnType is float64 and mName is rand_beta
returnType is int32 and mName is rand_cat
returnType is float64 and mName is rand_unif
returnType is float64 and mName is rand_beta
returnType is int32 and mName is rand_cat
```

will resume after 2:30pm

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala]: can be replicated by setting rand_cat's requirement based on the requiredness of the input PCArray, which is +PCArray[PFloat64]

**Comment by akotlar** [comment]
fixed!

File: hail/src/main/scala/is/hail/variant/HardCallView.scala [hunk 3]
```diff
@@ -83,19 +84,17 @@ final class HardCallView(rvType: PStruct, callField: String) {
   private val tgs = rvType.types(entriesIndex).asInstanceOf[PArray]
   private val tg = tgs.elementType.asInstanceOf[PStruct]
 
-  private def lookupField(name: String, expected: PType): (Boolean, Int) = {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/variant/HardCallView.scala:86]: function with exactly 1 use, unnecessary indirection

File: hail/src/main/scala/is/hail/expr/ir/package.scala [hunk 4]
```diff
@@ -84,6 +84,10 @@ package object ir {
     case None => fatal(s"no conversion found for $name(${args.map(_.typ).mkString(", ")}) => $rt")
   }
 
+  def invokeSeeded(name: String, rt: Type, args: IR*): IR = IRFunctionRegistry.lookupConversion(name, rt, args.init.map(_.typ)) match {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/package.scala:87]: will make future seeded tests easier / in line with other invoke uses.

**Pull request merged**